File: leetcode/679. 24 点游戏.go
func judgePoint24(cards []int) bool {
	nums := make([]float64, len(cards))
	for i, num := range cards {
		nums[i] = float64(num)
	}
	return dfs(nums)
}

func dfs(nums []float64) bool {
	if len(nums) == 0 {
		return false
	}
	if len(nums) == 1 {
		return math.Abs(nums[0]-24) < 1e-6
	}
	for i := 0; i < len(nums); i++ {
		for j := 0; j < len(nums); j++ {
			if i == j {
				continue
			}
			nextNums := make([]float64, 0, len(nums)-1)
			for k := 0; k < len(nums); k++ {
				if k != i && k != j {
					nextNums = append(nextNums, nums[k])
				}
			}
			for op := 0; op < 4; op++ {
				if op < 2 && i > j {
					continue
				}
				var res float64
				switch op {
				case 0:
					res = nums[i] + nums[j]
				case 1:
					res = nums[i] * nums[j]
				case 2:
					res = nums[i] - nums[j]
				case 3:
					if math.Abs(nums[j]) < 1e-6 {
						continue
					}
					res = nums[i] / nums[j]
				}
				nextNums = append(nextNums, res)
				if dfs(nextNums) {
					return true
				}
				nextNums = nextNums[:len(nextNums)-1]
			}
		}
	}
	return false
}

File: leetcode/679. 24 点游戏.rs
impl Solution {
    pub fn judge_point24(cards: Vec<i32>) -> bool {
        let nums: Vec<f64> = cards.iter().map(|&x| x as f64).collect();
        Self::dfs(&nums)
    }

    fn dfs(nums: &[f64]) -> bool {
        if nums.len() == 0 {
            return false;
        }
        if nums.len() == 1 {
            return (nums[0] - 24.0).abs() < 1e-6;
        }
        for i in 0..nums.len() {
            for j in 0..nums.len() {
                if i == j {
                    continue;
                }
                let mut next_nums: Vec<f64> = nums.iter()
                    .enumerate()
                    .filter(|&(k, _)| k != i && k != j)
                    .map(|(_, &x)| x)
                    .collect();
                for op in 0..4 {
                    // Skip commutative operations when i > j to avoid duplicates
                    if op < 2 && i > j {
                        continue;
                    }
                    
                    let res = match op {
                        0 => nums[i] + nums[j],
                        1 => nums[i] * nums[j],
                        2 => nums[i] - nums[j],
                        3 => {
                            if nums[j].abs() < 1e-6 {
                                continue;
                            }
                            nums[i] / nums[j]
                        }
                        _ => unreachable!(),
                    };
                    
                    next_nums.push(res);
                    if Self::dfs(&next_nums) {
                        return true;
                    }
                    next_nums.pop();
                }
            }
        }
        false
    }
}